aten/src/ATen/BatchedTensorImpl.cpp

```diff
@@ -0,0 +1,104 @@
+#include <ATen/BatchedTensorImpl.h>
+
+#include <ATen/WrapDimUtils.h>
+#include <c10/util/Exception.h>
+
+namespace at {
+
+BatchedTensorImpl::BatchedTensorImpl(Tensor value, BatchDims bdims)
+  : TensorImpl(
+      c10::DispatchKeySet(DispatchKey::Vmap),
+      value.dtype(),
+      value.device()
+    )
+  , value_(std::move(value))
+  , bdims_(std::move(bdims))
+{
+  TORCH_INTERNAL_ASSERT(value_.defined());
+
+  const auto public_dims = value_.dim() - bdims_.size();
+  const auto value_sizes = value_.sizes();
+  sizes_.clear();
+  sizes_.reserve(public_dims);
+  for (int64_t dim = 0; dim < public_dims; dim++) {
+    auto actual_dim = actualDim(dim, /*wrap_dim=*/false);
+    sizes_.push_back(value_sizes.at(actual_dim));
+  }
+  refresh_numel();
+}
+
+int64_t BatchedTensorImpl::actualDim(int64_t dim, bool wrap_dim) const {
+  if (wrap_dim) {
+    const auto ndim = sizes_.size();
+    dim = maybe_wrap_dim(dim, ndim);
+  }
+  auto is_bdim = createBatchDimBitset(bdims_);
+
+  // Example: assume dim = 3, and is_bdim = 10010011000...
+  // The 1's are batch dims and 0's are normal dims of the underlying value_ Tensor.
+  // actualDim gives us the index of `dim` in the `value_` Tensor, which is equivalent
+  // to asking "where does the 3rd (0-indexed) zero occur in the bitset?".
+  // The answer to that is index 5.
+  int64_t non_bdim_count = 0;
+  for (int64_t actual_dim = 0; actual_dim < kVmapMaxTensorDims; actual_dim++) {
```

Comment:
I feel like this should be implementable without a loop. Maybe https://stackoverflow.com/questions/7669057/find-nth-set-bit-in-an-int ? Don't know how portable the top liked answer is though.

Reply:
Although, since you're creating the bitset on the fly, I guess there isn't actually asymptotic improvement here :o)

Reply:
I don't know what CPUs we support, but the instruction set (BMI2) used in that answer should be on Intel CPUs newer than 6/2013 and AMD from newer than 2015. I'll leave a comment in the code as a TODO to figure that out